userspace/falco/application.h

```diff
@@ -134,6 +134,42 @@ class application {
 		bool proceed;
 	};
 
+	// Convenience method. Read a sequence of filenames and fill
+	// in a vector of rules contents.
+        // Also fill in the provided rules_contents_t with a mapping from
+        // filename (reference) to content (reference).
+	// falco_exception if any file could not be read.
+	template<class InputIterator>
+	void read_files(InputIterator begin, InputIterator end,
+			std::vector<std::string>& rules_contents,
+			falco::load_result::rules_contents_t& rc)
+	{
+		// Read the contents in a first pass
+		for(auto it = begin; it != end; it++)
+		{
+			std::string &filename = *it;
+			std::ifstream is;
+			is.open(filename);
+			if (!is.is_open())
+			{
+				throw falco_exception("Could not open file " + filename + " for reading");
+			}
+
+			std::string rules_content((istreambuf_iterator<char>(is)),
+						  istreambuf_iterator<char>());
+			rules_contents.emplace_back(std::move(rules_content));
+		}
+
+		// Populate the map in a second pass to avoid
+		// references becoming invalid.
+		uint32_t idx = 0;
+		for(auto it = begin; it != end; it++)
+		{
+			std::string &filename = *it;
+			rc.emplace(filename, rules_contents.at(idx++));
```

Comment:
```suggestion
			rc.emplace(filename, rules_contents.at(std::distance(begin, it)));
```
I think this might be safer here, what do you think?

In any case, why not range-iterating on `rules_contents` directly in this for loop? For begin and end, you have no real ordering guarantee because they are template iterators. I see this becoming quite fragile if we stop using vectors/lists for some reason.

Reply:
Sure, I can have two iterators for the names and contents. I'll also throw an exception if both iterators don't end up at their respective ends.